massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,1478 @@
+        let mut key = vec![KeyIndent::IsError as u8];
+        key.push(event.context.is_error as u8);
+        key.extend(self.get_event_key(event));
+        key
+    }
+
+    fn get_counter_key_from(&self, key: &[u8]) -> Vec<u8> {
+        vec![KeyIndent::Counter as u8, key[0]]
+    }
+
+    fn get_counter_key_from_indent(&self, indent: &KeyIndent) -> Vec<u8> {
+        vec![KeyIndent::Counter as u8, *indent as u8]
+    }
+
+    fn get_counter_key_bool_from(&self, key: &[u8], value: bool) -> Vec<u8> {
+        vec![u8::from(KeyIndent::Counter), key[0], u8::from(value)]
+    }
+
+    fn get_counter_key_bool_from_indent(&self, indent: &KeyIndent, value: bool) -> Vec<u8> {
+        vec![
+            u8::from(KeyIndent::Counter),
+            u8::from(*indent),
+            u8::from(value),
+        ]
+    }
+}
+
+pub(crate) struct EventCache {
+    /// RocksDB database
+    db: DB,
+    /// How many entries are in the db. Count is initialized at creation time by iterating
+    /// over all the entries in the db then it is maintained in memory
+    entry_count: usize,
+    /// Maximum number of entries we want to keep in the db.
+    /// When this maximum is reached `snip_amount` entries are removed
+    max_entry_count: usize,
+    /// How many entries are removed when `entry_count` reaches `max_entry_count`
+    snip_amount: usize,
+    /// Event serializer
+    event_ser: SCOutputEventSerializer,
+    /// Event deserializer
+    event_deser: SCOutputEventDeserializer,
+    /// Key builder
+    key_builder: EventCacheKeyBuilder,
+    /// First event slot in db
+    first_slot: Slot,
+    /// Last event slot in db
+    last_slot: Slot,
+    /// Thread count
+    thread_count: u8,
+    /// Maximum number of events per operation
+    max_events_per_operation: u64,
+    /// Maximum number of operations per block
+    max_operations_per_block: u64,
+}
+
+impl EventCache {
+    #[allow(clippy::too_many_arguments)]
+    /// Create a new EventCache
+    pub fn new(
+        path: &Path,
+        max_entry_count: usize,
+        snip_amount: usize,
+        thread_count: u8,
+        max_recursive_call_depth: u16,
+        max_event_data_length: u64,
+        max_events_per_operation: u64,
+        max_operations_per_block: u64,
+    ) -> Self {
+        // Clear the db
+        DB::destroy(&Options::default(), path).expect(DESTROY_ERROR);
+        let options = {
+            let mut opts = Options::default();
+            opts.create_if_missing(true);
+            opts.set_merge_operator_associative("counter merge operator", counter_merge);
+            opts
+        };
+        let db = DB::open(&options, path).expect(OPEN_ERROR);
+
+        let key_builder = EventCacheKeyBuilder::new();
+        // init counters
+        let mut batch = WriteBatch::default();
+        let value = 0u64.to_be_bytes();
+        let key_counter = key_builder.get_counter_key_from_indent(&KeyIndent::EmitterAddress);
```

Comment:
we don't need to count the number of elements in the general emitter address index. What we need is to know how many emitter address index items there are FOR A GIVEN EMITTER ADDRESS. That's how we will more accurately determine which cache to use as pivot.

For example if someone queries an is_error=true + an emitter address I would:
* get the number of events in the emitter address index for that particular address
* get the number of events in the is_error index for which the value is true
* then query firs the one that has less elements and filter over the other criterion

Reply:
Done